clients/src/main/java/org/apache/kafka/clients/consumer/internals/GroupCoordinator.java

```diff
@@ -0,0 +1,635 @@
+                // re-discover the coordinator and retry with backoff
+                coordinatorDead();
+                log.info("Attempt to join group {} failed due to obsolete coordinator information, retrying.",
+                        groupId);
+                future.raise(Errors.forCode(errorCode));
+            } else if (errorCode == Errors.INCONSISTENT_GROUP_PROTOCOL.code()
+                    || errorCode == Errors.INVALID_SESSION_TIMEOUT.code()) {
+                // log the error and re-throw the exception
+                Errors error = Errors.forCode(errorCode);
+                log.error("Attempt to join group {} failed due to: {}",
+                        groupId, error.exception().getMessage());
+                future.raise(error);
+            } else {
+                // unexpected error, throw the exception
+                future.raise(new KafkaException("Unexpected error in join group response: "
+                        + Errors.forCode(joinResponse.errorCode()).exception().getMessage()));
+            }
+        }
+    }
+
+    private RequestFuture<S> performSync(JoinGroupResponse joinResponse) {
+        if (joinResponse.isLeader()) {
+            try {
+                // perform the leader synchronization and send back the assignment for the group
+                Map<String, M> metadata = deserializeMetadata(joinResponse.members());
+                Map<String, S> groupAssignment = doSync(joinResponse.leaderId(), joinResponse.subProtocol(), metadata);
+
+                SyncGroupRequest request = new SyncGroupRequest(groupId, generation,
+                        memberId, serializeAssignment(groupAssignment));
+                log.debug("Issuing leader SyncGroup ({}: {}) to coordinator {}", ApiKeys.SYNC_GROUP, request, this.coordinator.id());
+                return sendSyncGroupRequest(request);
+            } catch (RuntimeException e) {
+                return RequestFuture.failure(e);
+            }
+        } else {
+            // send follower's sync group with an empty assignment
+            SyncGroupRequest request = new SyncGroupRequest(groupId, generation,
+                    memberId, Collections.<String, ByteBuffer>emptyMap());
+            log.debug("Issuing follower SyncGroup ({}: {}) to coordinator {}", ApiKeys.SYNC_GROUP, request, this.coordinator.id());
+            return sendSyncGroupRequest(request);
+        }
+    }
+
+    private RequestFuture<S> sendSyncGroupRequest(SyncGroupRequest request) {
+        if (coordinatorUnknown())
+            return RequestFuture.coordinatorNotAvailable();
+        return client.send(coordinator, ApiKeys.SYNC_GROUP, request)
+                .compose(new SyncGroupRequestHandler());
+    }
+
+    private class SyncGroupRequestHandler extends CoordinatorResponseHandler<SyncGroupResponse, S> {
+
+        @Override
+        public SyncGroupResponse parse(ClientResponse response) {
```

Comment:
I'm wondering if we can put the parse() function in a centralized place since their implementations are all similar?